### PR TITLE
Address problems with use_deterministic_cudnn test decorator

### PR DESCRIPTION
**Pull request by duncanriach**
There is something not right here, and I'm trying to figure out how best to fix it.

`use_deterministic_cudnn` in `tensorflow/python/framework/test_util.py` is a test decorator that was added in [this commit](https://github.com/tensorflow/tensorflow/commit/c27909ea80e8823dbf4f7176ab69991a630356a1) on 2019-03-13.

The only place it's used is to decorate `layer_test` in `tensorflow/python/keras/testing_utils.py`, which is called (only) from a bunch of test files in `tensorflow/python/keras/layers`.

I initially noticed that the docstring for `use_deterministic_cudnn` does not match what the decorator actually does, and I intended to fix that. The decorator sets `TF_CUDNN_DETERMINISTIC=true`, and that not only disables cuDNN auto-tuning (from an end-user perspective), but also ensures the deterministic selection of deterministic algorithms for the back-prop of cuDNN convolution and max-pooling.

But the problems go deeper. This decorator intends to enable `TF_CUDNN_DETERMINISTIC` only for the decorated test. However, `TF_CUDNN_DETERMINISTIC` is currently implemented so that its value is cached in a static variable the first time it's used. This means that the implied promise of the decorator code to restore the `TF_CUDNN_DETERMINISTIC` setting cannot be fulfilled. Worse still, if an earlier-running test causes `TF_CUDNN_DETERMINISTIC` to be cached (as "0" or "false") then the decorator will have no effect whatsoever.

Additionally, the underlying `TF_CUDNN_DETERMINISTIC` functionality is currently implemented in a way that leads to the caching of the chosen algorithms (for any given layer configuration), even if the environment variable itself were not cached.

In this current pull request, I'm proposing changing the decorator to use `TF_CUDNN_USE_AUTOTUNE=false` (instead of `TF_CUDNN_DETERMINISTIC=true`), which both makes the code match the docstring and works in terms of environment variable caching: I believe that `TF_CUDNN_USE_AUTOTUNE` is freshly re-evaluated every time a cuDNN convolution is launched.

Whether this proposed change is appropriate or not depends on what is meant by "Some tests want to base assertions on a graph being _isomorphic with a copy_." (from the docstring). What is meant by the graph being isomorphic with a copy? Is this requiring bit-exact determinism from the cuDNN convolution and max-pooling back-prop algorithms? If yes, then this per-test-decorator-based solution is probably not appropriate and `TF_CUDNN_DETERMINISTIC` should be set to "true" in a broader scope. An example approach is shown [here](https://github.com/tensorflow/tensorflow/blob/v2.0.0/tensorflow/python/kernel_tests/cudnn_determinism_test.py#L102).

I would like to work with the right person/people to help resolve this issue.

### CODE REVIEW

**Comment by jaingaurav** [comment]
@chsigg: You might be a better person for this PR

**Comment by tpopp** [comment]
> What is meant by the graph being isomorphic with a copy?

This is poorly phrased, but it just refers to the fact that inside of layer_test, there is a test that copying the model and running the model returns approximately the same result. With a different auto-tuning algorithm, the result could change by a large amount that caused some tests to fail.

> Is this requiring bit-exact determinism from the cuDNN convolution and max-pooling back-prop algorithms? If yes, then this per-test-decorator-based solution is probably not appropriate and TF_CUDNN_DETERMINISTIC should be set to "true" in a broader scope.

It's requiring determinism of the algorithm chosen to limit variance (though at the time still not bit-exact determinism probably because not all reductions are deterministic for example), during the execution of all graphs during this function, relative to each other, but not relative to some baseline across executions of the test case.

Tensorflow has begun to have a policy with 2.0 not to use environment variables for configuring tests. In this case, it's okay to fix a bug without updating it to the new way of doing things, but the new ideal way of doing this would be similar to what is done here: https://github.com/tensorflow/tensorflow/commit/91a5a7318f77a7e6c7198d12b20c5414de29664d with TF_SetTfXlaCpuGlobalJit. It's also not ideal to rely on the implementation not caching the environment variable, when this optimization has been done in other areas, but for a test that's mostly fine. There is one blocker issue though, which is that XLA runs of this test would become flaky because the TF:XLA bridge does not know of TF_CUDNN_USE_AUTOTUNE; this is a shortcoming (bug) in the tf-xla bridge that it didn't know of this switch.

I'd rather avoid the broader scope setting of a variable. While it is more proper than what is there right now and will have less surprises, it might lead people to having bad tests when they are creating additional test cases in the same file. It also doesn't fix the issue with `layer_test` because that function might be imported and called by many different test files. 

I'm giving you only a partial answer now, so you know you're not being ignored. I'll see tomorrow if it's trivial to add bridge support for TF_CUDNN_USE_ATOTUNE. The alternative that is more appropriate but would be more work for you is if you're willing to make the TF_CUDNN_DETERMINISTIC a variable that can be programatically set rather than just static, is initially set with the environment variable, and then adding the TF_CAPI_EXPORT related code like in the linked change.

**Comment by duncanriach** [comment]
Hi @tpopp, Thanks so much for such a comprehensive reply.

You've provided some very compelling arguments for controlling this in the decorator used by `layer_test`, rather than by setting a variable at a larger scope. I agree with you on that.

Even though XLA currently doesn't know about `TF_CUDNN_DETERMINISTIC` explicitly, I think that `TF_CUDNN_DETERMINISTIC` is currently working in this test decorator (on XLA) because of the way that `TF_CUDNN_DETERMINISTIC` functionality is currently implemented. `TF_CUDNN_DETERMINISTIC` currently works (in `tensorflow/stream_executor/cuda/cuda_dnn.cc`) by restricting the forward and two backward algorithms available to both the XLA and non-XLA auto-tuning code to a single, deterministic algorithm in each case, via the whitelists of algorithms returned by `CudnnSupport::GetConvolve*Algorithms()` in the stream-executor.

I'm about to issue a pull request for a bug-fix for `TF_CUDNN_DETERMINISTIC` that requires those returned whitelists to contain all deterministic algorithms and for the auto-tuning code (both XLA and non-XLA) to operate deterministically when `TF_CUDNN_DETERMINISTIC=true`. At this point, I have not implemented that using the `TF_CAPI_EXPORT` mechanism, which I had not known about. The way I have implemented it is that the XLA auto-tuning code just calls a function in the main TF object, which holds the cached setting of `TF_CUDNN_DETERMINISTIC`.

With the current implementation of `TF_CUDNN_DETERMINISTIC`, with the upcoming bug fix, and also with a later, higher-performance version that I'm planning, caching of cuDNN algorithm selection is part of the default functionality of the deterministic mode. This means that adding the ability to enable and disable it on a layer-by-layer basis would require additional algorithm-cache invalidation functionality (at least for the current version and the upcoming bug-fix version) which, as far as I can tell, would only be useful for the testing case that is the focus of this current PR.

Given that the default functionality of `TF_CUDNN_USE_AUTOTUNE` lends itself to being enabled and disabled on-the-fly (and this is how it's currently implemented for non-XLA) and given that the `TF_CUDNN_AUTOTUNE` environment variable is also currently not cached, I think that it's preferable that we use `TF_CUDNN_AUTOTUNE`.

If you agree, I could have a go at adding `TF_CUDNN_AUTOTUNE` functionality to the XLA auto-tune code in `tensorflow/compiler/xla/service/gpu/gpu_conv_algorithm_picker.cc` and pass the setting through from the environment variable using the `TF_CAPI_EXPORT` mechanism (as you suggested in your above comment for `TF_CUDNN_DETERMINISTIC`).

**Comment by tpopp** [comment]
Hi @duncanriach,

I'm very sorry for not responding. I don't do PRs normally, so I completely forgot.

Either of your suggestions in the last 2 paragraphs is good in my opinion, and I will sign off on either as long as it would not result in broken tests running with XLA (Running the test with TF_XLA_FLAGS="--tf_xla_auto_jit=2 --tf_xla_cpu_global_jit --tf_xla_min_cluster_size=1" should be enough to know).

Given this decorator is only used for a few tests, and the use of autotuning in general for these tests is debatable (both sides having valid points), I think any forward progress is good, and we shouldn't stress too much (but you're right about the current scheme being completely broken and us being lucky).

Again, I apologize for ignoring you, and I'll do better about checking github.

**Comment by duncanriach** [comment]
No problem, @tpopp. Thanks for reviewing this PR.
I have some other things that I'm more urgently working on and I will return to this PR in the next few weeks and decide how to proceed then.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by duncanriach** [comment]
The direction of this PR is dependent on the outcome of [PR 34951](https://github.com/tensorflow/tensorflow/pull/34951), which I'm also dealing with at a higher priority than this one, and which is currently waiting for a review response.

**Comment by duncanriach** [comment]
I'm going to close this pull request and open it again when I have the time and clarity to push it through to a merge.

**Comment by duncanriach** [comment]
I'm re-opening this pull-request so that I can submit some additional changes for review, discussion, and potential merge.

**Comment by duncanriach** [comment]
@tpopp, I have discovered that the XLA code does already have a mechanism for disabling cuDNN auto-tuning. Auto-tuning is disabled by adding `--xla_gpu_disable_autotune` to `XLA_FLAGS`. It takes effect in [xla/.../gpu_conv_algorithm_picker.cc](https://github.com/tensorflow/tensorflow/blob/240497c2b2058ed25b09308dd906441882ca7611/tensorflow/compiler/xla/service/gpu/gpu_conv_algorithm_picker.cc#L721) and in [xla/.../gemm_algorithm_picker.cc](https://github.com/tensorflow/tensorflow/blob/3b71976900cb1bdb7016441076c70dcc3296b8b3/tensorflow/compiler/xla/service/gpu/gemm_algorithm_picker.cc#L319).

I have submitted an additional commit to this current PR which controls this switch from the `disable_cudnn_autotune` decorator. I have tested this locally only using `//tensorflow/python/keras/convolutional_test` both with and without `TF_XLA_FLAGS="--tf_xla_auto_jit=2 --tf_xla_cpu_global_jit --tf_xla_min_cluster_size=1"` These tests pass.

It's not totally clear to me that `XLA_FLAGS` changes will take effect between test-cases, as this fix expects and implies.

If this current solution is **not** acceptable to you, then another option might be to add an equivalent `pywrap_tensorflow` switch and then get it to that same location in `xla/.../gpu_conv_algorithm_picker.cc` and `xla/.../gemm_algorithm_picker.cc`.

**Comment by duncanriach** [comment]
@gbaned please will you remove the following tags: `stalled` and `stat:awaiting response`. This PR is now waiting for review from @tpopp.

**Comment by duncanriach** [comment]
Adding reference to [PR 35621](https://github.com/tensorflow/tensorflow/pull/35621) (Replace --xla_gpu_disable_autotune option with --xla_gpu_autotune_level) which intends to change an XLA flag that I have proposed to utilize in this current PR.

**Review by tpopp — APPROVED** [review]

**Comment by tpopp** [comment]
Unless it is hidden further down, it looks like the flags will be parsed multiple times for XLA_FLAGS, so I'm going to approve this.

File: tensorflow/python/framework/test_util.py
Comment (bas-aarts) [inline tensorflow/python/framework/test_util.py]: Assuming that the decorator takes precedence over XLA_FLAGS, new_xla_flags should be appended in stead of pre-pended. With the current change, if XLA_FLAGS contains --xla_gpu_disable_autotune=false, the decorator does not achieve its intended behavior.
If my assumption is incorrect, pre-pending is the wat to go

Comment (bas-aarts) [reply tensorflow/python/framework/test_util.py]: Since this change was merged, I'll rebase [PR 35621](https://github.com/tensorflow/tensorflow/pull/35621)  and make the change.

Comment (duncanriach) [reply tensorflow/python/framework/test_util.py]: Well spotted and reported, @bas-aarts. I didn't see your comment in time to address it prior to the merge. Thank you for taking care of this in your PR.

**Pull request merged**
